[infra] Shorten IDs for ML jobs

miltonhultgren opened the pull request:
Closes #47477

### Summary

ML job IDs have a limit of 64 characters. For the log ML jobs we add the string `kibana-logs-ui` plus the space and log view IDs as a prefix to the job names (`log-entry-rate` and `log-entry-categories-count`) which can quickly eat up the 64 character limit (even our own Stack Monitoring log view hits the limit). This prevents users from being able to create ML jobs and it's hard to rename a space or log view, and the limit is not hinted at during space creation (because they are unrelated in some sense).

In order to achieve a more stable length to the ID, this PR introduces a new format for the prefix which creates a UUID v5 which uses the space and log view ID as seed information (it then removes the dashes to still be within the size limit for the categorization job).

Since there is no technical difference between the new and old format, this PR makes an effort to continue to support the old format and allow migration of old jobs as needed. The old jobs work and may contain important data so the user should not feel forced to migrate.

The main addition is a new small API that checks if any ML jobs are available and which format they use for the ID so that the app can request data accordingly and the APIs have been modified to take the ID format into account (except during creation which should always use the new format).

The solution applied is not ideal. It simply passes the ID format along with the space and log view ID to each point where the ID is re-created (which is multiple). The ideal solution would be to store the job data in the store and pass that around instead but that seemed like a considerably larger effort. This PR does introduce some functional tests around the ML job creation process, so such a future refactor should be a bit safer than previously. 

### How to test

* Start from `main`
* Start Elasticsearch
* Start Kibana
* Load the Sample web logs (Kibana home -> Try sample data -> Other sample data sets)
* Visit the Anomalies page in the Logs UI
* Set up any of the two ML jobs or both, wait for some results to show up
* Checkout the PR branch
* Visit the anomalies page and verify that it still works (requests go to resolve the ID format, should return 'legacy' which should then load the data for the legacy job)
* Recreate the ML job and verify that the new job works and results still show up (new requests should go out with the new format being used, which may be a mixed mode if you have two jobs and only migrate one of them)

miltonhultgren committed:
[infra] Shorten IDs for ML jobs (#47477)

apmmachine commented:
### :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

Just comment with:
- `/oblt-deploy` : Deploy a Kibana instance using the Observability test environments.
- `/oblt-deploy-serverless` : Deploy a serverless Kibana instance using the Observability test environments.
- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)

</p>
</details>

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Use uuid.v5 instead of crypto, update description of created jobs

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Add basic tests and test helpers

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into 47477-shorten-ml-job-ids

miltonhultgren committed:
Add more tests

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into 47477-shorten-ml-job-ids

miltonhultgren committed:
Add more tests

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren commented on x-pack/plugins/infra/common/log_analysis/job_parameters.ts:
```diff
@@ -6,21 +6,39 @@
+  if (idFormat === 'legacy') {
+    return `kibana-logs-ui-${spaceId}-${sourceId}-`;
+  } else {
+    const uuid = v5(`${spaceId}-${sourceId}`, ID_NAMESPACE).replaceAll('-', '');
```
Add comment about why it's cutting out the dashes

miltonhultgren committed:
Add comment about removing dashes from UUID

elasticmachine commented:
Pinging @elastic/infra-monitoring-ui (Team:Infra Monitoring UI)

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren commented:
In what situation would a ML job be missing the saved object? I'm guessing for jobs that were created before they were space aware so we could safely assume they're all in the old format and this would happen as part of an upgrade?
Or perhaps it may happen if the user deleted the related saved objects (for some reason)?

In the new format we don't directly add the space ID to the ID, though it is part of the text we hash.

@weltenwort What's your take here? With the new hash format we cannot deduce the space name, at the same time ML jobs should be space aware unless tampered with so I'm not sure we need to support this anymore. 
Perhaps it would be acceptable that jobs in the new format end up in the wrong space if the user manually deletes a saved object?

jgowdyelastic commented:
>In what situation would a ML job be missing the saved object?

If the job has been created directly in elasticsearch, which is very unlikely for these logs jobs.
ML's saved objects are not listed in the management page, and cannot be manually deleted.

>I'm guessing for jobs that were created before they were space aware

Yes, I believe that is exactly why we added this logic.

weltenwort commented:
Since any new jobs will have saved objects the space-awareness migration should not be necessary for them. That means we just have to make sure the migration code ignores them and doesn't break.

@jgowdyelastic we're including the space id in the hash to avoid collisions between concurrently running jobs in different spaces. Or does the ML plugin perform some kind of collision avoidance anyway?

jgowdyelastic commented:
> @jgowdyelastic we're including the space id in the hash to avoid collisions between concurrently running jobs in different spaces. Or does the ML plugin perform some kind of collision avoidance anyway?

Ah ok, as there is no concept of spaces in es, we cannot allow duplicate IDs even if they are in separate spaces. So you'll need to keep this UUID and based on the space.

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

crespocarlos commented on x-pack/plugins/infra/public/pages/logs/log_entry_rate/page_providers.tsx:
```diff
@@ -29,16 +31,19 @@ export const LogEntryRatePageProviders: React.FunctionComponent = ({ children })
   } else if (!isPersistedLogView) {
     return <InlineLogViewSplashPage revertToDefaultLogView={revertToDefaultLogView} />;
-  } else if (isLoading || isUninitialized) {
+  } else if (isLoading || isUninitialized || isLoadingLogAnalysisIdFormats) {
```
We could remove the `idFormats!` below by doing this: 

```suggestion
  } else if (isLoading || isUninitialized || isLoadingLogAnalysisIdFormats || !idFormats) {
```

crespocarlos commented on x-pack/plugins/infra/public/pages/logs/log_entry_rate/sections/anomalies/expanded_row.tsx:
```diff
@@ -44,6 +50,7 @@ export const AnomaliesTableExpandedRow: React.FunctionComponent<{
     endTime: anomaly.startTime + anomaly.duration,
     exampleCount: EXAMPLE_COUNT,
     logViewReference,
+    idFormat: idFormats![logEntryRateJobType],
```
Here too: `useLogEntryExamples` could handle when `idFormats` is `undefined`

crespocarlos commented on x-pack/plugins/infra/public/pages/logs/shared/use_log_ml_job_id_formats_shim.tsx:
```diff
@@ -0,0 +1,85 @@
+    getLogAnalysisIdFormats();
+  }, [getLogAnalysisIdFormats]);
+
+  const isLoadingLogAnalysisIdFormats = useMemo(
+    () => getLogAnalysisIdFormatsRequest.state === 'pending',
+    [getLogAnalysisIdFormatsRequest.state]
+  );
+
+  const hasFailedLoadingLogAnalysisIdFormats = useMemo(
+    () => getLogAnalysisIdFormatsRequest.state === 'rejected',
+    [getLogAnalysisIdFormatsRequest.state]
+  );
```
Q: Do we need to `memoize` these?

crespocarlos commented on x-pack/plugins/infra/public/pages/logs/shared/use_log_ml_job_id_formats_shim.tsx:
```diff
@@ -0,0 +1,85 @@
+    [getLogAnalysisIdFormatsRequest.state]
+  );
+
+  const migrateIdFormat = useCallback(
+    (jobType: JobType) => {
+      setIdFormats({
+        ...idFormats!,
+        [jobType]: 'hashed',
+      });
+    },
+    [idFormats]
+  );
```
Would this work?

```suggestion
 const migrateIdFormat = useCallback((jobType: JobType) => {
    setIdFormats(
      (value) => ({ ...(value ? value : {}), [jobType]: 'hashed' } as IdFormatByJobType)
    );
  }, []);
```

miltonhultgren replied:
That wouldn't work like that because it only migrates one job type at a time and the type expects values for both job types to be present (which it should be if it has actually loaded the formats from the API).
But I'll do something similar which is to just keep setting it to null if it hasn't loaded it from the API yet which I think would be more correct.

crespocarlos commented on x-pack/plugins/infra/public/containers/logs/log_analysis/api/ml_cleanup.ts:
```diff
@@ -8,28 +8,30 @@
 import * as rt from 'io-ts';
 import type { HttpHandler } from '@kbn/core/public';
 
+import { IdFormat, JobType } from '../../../../../common/http_api/log_analysis/v1/id_formats';
```
There are many imports referring to `http_api/log_analysis/v1`. We could centralize them in [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/infra/common/http_api/latest.ts) and replace all references to `common/http_api/log_analysis/v1` with `common/http_api/log_analysis/latest`. This would help maintain this feature should the version changes in the future

miltonhultgren replied:
Did this, but didn't touch `/common` since the latest folder depends on the other http_api files

crespocarlos commented on x-pack/plugins/infra/common/log_analysis/job_parameters.ts:
```diff
@@ -6,21 +6,41 @@
+  } else {
+    // A UUID is 36 characters but based on the ML job names for logs, our limit is 32 characters
+    // Thus we remove the 4 dashes
+    const uuid = v5(`${spaceId}-${sourceId}`, ID_NAMESPACE).replaceAll('-', '');
```
Suggestion: You could use `uuidv5.DNS` instead of `ID_NAMESPACE`

miltonhultgren replied:
Since we're not hashing a URL/domain name that might be confusing and it's not a big overhead to use a custom one, I'll stick with that.

crespocarlos reviewed (commented):
Thanks for this change. 

I haven't had the chance to test the feature yet, but I've left some comments while reading the PR.

elasticmachine commented:
Pinging @elastic/obs-ux-logs-team (Team:obs-ux-logs)

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into 47477-shorten-ml-job-ids

miltonhultgren committed:
Change import to refernece http_api/latest

miltonhultgren committed:
Remove unnecessary useMemo calls, use previous value in state update and add guard

miltonhultgren committed:
Shuffle around null checks

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren commented:
@crespocarlos Thanks for the code review! I think I've addressed your comments.

I made the change for the null checking for the `idFormats` but I'm not 100% sure it's needed since the check happening inside the page_providers block all the other code from executing. Though, if that where to change it's good to have those null checks in place so we get a good error message out instead!

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/eslint --no-cache --fix'

crespocarlos reviewed (approved):
LGTM.
Thanks for applying the suggestions. Both `job_id` and `datafeed_id` have the new format.

<img width="572" alt="image" src="https://github.com/elastic/kibana/assets/2767137/22668ae2-963b-4542-9b5d-332082becde5">

It works according to the steps in the description.

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' of github.com:elastic/kibana into 47477-shorten-ml-job-ids

miltonhultgren committed:
Move tests to new location

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

miltonhultgren committed:
Merge branch 'main' into 47477-shorten-ml-job-ids

tonyghiani commented on x-pack/plugins/infra/common/log_analysis/job_parameters.ts:
```diff
@@ -6,21 +6,41 @@
 
-export const getJobIdPrefix = (spaceId: string, sourceId: string) =>
-  `kibana-logs-ui-${spaceId}-${sourceId}-`;
+const ID_NAMESPACE = 'f91b78c0-fdd3-425d-a4ba-4c028fe57e0f';
```
**nit:** Hardcoded IDs in the codebase are a bit of a code smell in the long run, can we just comment on what it represents and why can we keep it hardcoded? Future ourselves will thank us for this 😄

miltonhultgren replied:
The auto merge went through but to address your concern, I feel the variable name and it's usage is clear enough? 

We need to generate a unique ID for each job so we use `uuid.v5` which accepts a UUID to use as the namespace for the hashing and a name unique in that namespace.

tonyghiani reviewed (approved):
LGTM, everything requested change is already addressed and I didn't find any issue, the reproduction works as explained in the description, just left a (non-blocking) nit, thanks for this work!
